src/box/lua/key_def.c

```diff
@@ -397,13 +415,14 @@ lbox_key_def_compare_with_key(struct lua_State *L)
  * Raise error otherwise.
  */
 static int
-lbox_key_def_merge(struct lua_State *L)
+lbox_key_def_merge(struct lua_State *L, int idx, const char *obj_name)
 {
-	struct key_def *key_def_a, *key_def_b;
-	if (lua_gettop(L) != 2 ||
-	    (key_def_a = luaT_check_key_def(L, 1)) == NULL ||
-	    (key_def_b = luaT_check_key_def(L, 2)) == NULL)
-		return luaL_error(L, "Usage: key_def:merge(second_key_def)");
+	struct key_def *key_def_a = luaT_check_key_def(L, idx);
+	struct key_def *key_def_b = luaT_check_key_def(L, 2);
+	if (lua_gettop(L) != 2 || key_def_a == NULL || key_def_b == NULL) {
+		return luaL_error(L, "Usage: %s:merge(second_key_def)",
+				  obj_name);
+	}
```

Comment:
It looks lopsided:

1. `key_def_a` is at `idx`, OK.
2. `key_def_b` is at `2`, hm.
3. The stack top is strictly `2`.

So it looks like the function is to be called only from Lua (I mean, wrapped and registered with lua_pushcfunction()), otherwise the stack top condition is strange. And so it is not clear why `key_def_a` is not always at slot `1`.

Reply:
The question about the stack top checks is applicable to other functions as well.

Reply:
Now all these methods can be called in two ways:

1.
```lua
key_def = require('key_def')
k = key_def.new(idx.parts)
k:merge(second_key_def)
```
Here self is key_def_a (slot 1), key_def_b is at 2.

2.
```lua
idx.parts:merge(second_key_def)
```
Here unused self is at slot 1, key_def_b is at 2, key_def_a is in the upvalue.

Reply:
If we'll go with this implementation, it should be commented somehow in the code, because it is not obvious at reading. It is a part of the function contract in fact.

Reply:
Added a comment.

Reply:
Irrelevant now. Closing.